internal/querynode/result.go

```diff
@@ -294,13 +294,19 @@ func mergeInternalRetrieveResult(ctx context.Context, retrieveResults []*interna
 		}
 
 		pk := typeutil.GetPK(validRetrieveResults[sel].GetIds(), cursors[sel])
-		if _, ok := idSet[pk]; !ok {
+		ts := typeutil.GetTS(validRetrieveResults[sel], cursors[sel])
+		if _, ok := idTsMap[pk]; !ok {
 			typeutil.AppendPKs(ret.Ids, pk)
 			typeutil.AppendFieldData(ret.FieldsData, validRetrieveResults[sel].GetFieldsData(), cursors[sel])
-			idSet[pk] = struct{}{}
+			idTsMap[pk] = ts
 		} else {
 			// primary keys duplicate
 			skipDupCnt++
+			if ts != 0 && ts > idTsMap[pk] {
+				idTsMap[pk] = ts
+				ret.FieldsData = ret.FieldsData[:len(ret.FieldsData)-1]
```

Comment:
FieldsData is a list of fields, you need to operate on the separate field in the FieldsData